enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,486 @@
+		})
+		if err != nil {
+			return xerrors.Errorf("insert workspace: %w", err)
+		}
+
+		// We have to refetch the workspace for the joined in fields.
+		workspace, err := db.GetWorkspaceByID(ctx, minimumWorkspace.ID)
+		if err != nil {
+			return xerrors.Errorf("get workspace by ID: %w", err)
+		}
+
+		c.logger.Info(ctx, "attempting to create prebuild", slog.F("name", name),
+			slog.F("workspace_id", prebuildID.String()), slog.F("preset_id", presetID.String()))
+
+		return c.provision(ctx, db, prebuildID, template, presetID, database.WorkspaceTransitionStart, workspace)
+	}, &database.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  false,
+	})
+}
+
+func (c *StoreReconciler) deletePrebuild(ctx context.Context, prebuildID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
+	return c.store.InTx(func(db database.Store) error {
+		workspace, err := db.GetWorkspaceByID(ctx, prebuildID)
+		if err != nil {
+			return xerrors.Errorf("get workspace by ID: %w", err)
+		}
+
+		template, err := db.GetTemplateByID(ctx, templateID)
+		if err != nil {
+			return xerrors.Errorf("failed to get template: %w", err)
+		}
+
+		c.logger.Info(ctx, "attempting to delete prebuild",
+			slog.F("workspace_id", prebuildID.String()), slog.F("preset_id", presetID.String()))
+
+		return c.provision(ctx, db, prebuildID, template, presetID, database.WorkspaceTransitionDelete, workspace)
+	}, &database.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  false,
+	})
+}
+
+func (c *StoreReconciler) provision(ctx context.Context, db database.Store, prebuildID uuid.UUID, template database.Template, presetID uuid.UUID, transition database.WorkspaceTransition, workspace database.Workspace) error {
+	tvp, err := db.GetPresetParametersByTemplateVersionID(ctx, template.ActiveVersionID)
+	if err != nil {
+		return xerrors.Errorf("fetch preset details: %w", err)
+	}
+
+	var params []codersdk.WorkspaceBuildParameter
+	for _, param := range tvp {
+		// TODO: don't fetch in the first place.
+		if param.TemplateVersionPresetID != presetID {
+			continue
+		}
+
+		params = append(params, codersdk.WorkspaceBuildParameter{
+			Name:  param.Name,
+			Value: param.Value,
+		})
+	}
+
+	builder := wsbuilder.New(workspace, transition).
+		Reason(database.BuildReasonInitiator).
+		Initiator(prebuilds.SystemUserID).
+		ActiveVersion().
+		VersionID(template.ActiveVersionID).
+		Prebuild().
+		TemplateVersionPresetID(presetID)
+
+	// We only inject the required params when the prebuild is being created.
+	// This mirrors the behavior of regular workspace deletion (see cli/delete.go).
+	if transition != database.WorkspaceTransitionDelete {
+		builder = builder.RichParameterValues(params)
+	}
+
+	_, provisionerJob, _, err := builder.Build(
+		ctx,
+		db,
+		func(action policy.Action, object rbac.Objecter) bool {
+			return true // TODO: harden?
+		},
+		audit.WorkspaceBuildBaggage{},
+	)
+	if err != nil {
+		return xerrors.Errorf("provision workspace: %w", err)
+	}
```

Comment:
Looking at the code coverage, I believe we're missing a failure test for this. One possible scenario is that a preset is defined for which a required parameter value is not specified, which means when the prebuild attempts to build it'll fail immediately. This might be a bit of a flaw in our control-flow right now because I believe it'll keep retrying without backing off every interval. This might be OK since it'll log the error and operators will quickly see what is the matter, although I don't think we expose any metric for this case at the moment.

Reply:
@evgeniy-scherbina I believe you said this one was still being worked on, correct?

Reply:
@dannykopping I added simple test: https://github.com/coder/coder/pull/17261/commits/62fb3f4c184a105840cab0c844c9d04b2f3c7696